### PR TITLE
[SYCL] Enable device library online link

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: jinge90 <ge.jin@intel.com>

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, all.
Please ignore this pr, we create it just for some local experiments.

Thanks very much.

**Comment by AGindinson** [comment]
@jinge90, thanks for the clarification! Just as an FYI, the [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality can be used for work-in-progress/experimental cases. Later on, if the PR is decided to be good for review, it can always be converted to a "full-scale" PR.

**Comment by jinge90** [comment]
Hi, @capatober 
Please try this branch. And please "export SYCL_DEVICE_FILTER" to gpu opencl backend in your experiment.
Thanks very much.

**Comment by jinge90** [comment]
SYCL :: DeprecatedFeatures/SpecConsts1.2.1/spec_const_redefine.cpp
  SYCL :: DeprecatedFeatures/basic.cpp
  SYCL :: DeprecatedFeatures/spec_consts.cpp
  SYCL :: ESIMD/sycl_esimd_mix.cpp
  SYCL :: KernelAndProgram/build-log.cpp
  SYCL :: KernelAndProgram/cache_env_vars.cpp
  SYCL :: KernelAndProgram/cache_env_vars_lin.cpp
  SYCL :: Regression/implicit_kernel_bundle_image_filtering.cpp
Failed and we may need to update those tests. Will investigate case by case.

**Comment by jinge90** [comment]
One devicelib test cmath_test failed on OCL CPU device, suspect it is a OCL CPU runtime bug.

**Comment by jinge90** [comment]
Hi, @gmlueck and @smaslov-intel 
This PR is for re-enable devicelib online link.

Thanks very much.

File: llvm/tools/sycl-post-link/SYCLDeviceLibReqMask.cpp [hunk 1]
```diff
@@ -198,11 +200,7 @@ uint32_t getModuleDeviceLibReqMask(const Module &M) {
   // Device libraries will be enabled only for spir-v module.
   if (!llvm::Triple(M.getTargetTriple()).isSPIR())
     return 0;
-  // 0x1 means sycl runtime will link and load libsycl-fallback-assert.spv as
-  // default. In fact, default link assert spv is not necessary but dramatic
-  // perf regression is observed if we don't link any device library. The perf
-  // regression is caused by a clang issue.
-  uint32_t ReqMask = 0x1;
+  uint32_t ReqMask = 0;
```
Comment (gmlueck) [inline llvm/tools/sycl-post-link/SYCLDeviceLibReqMask.cpp:203]: I'm just curious, what was the performance issue described in the old comment?  Is that fixed now?

Comment (jinge90) [reply llvm/tools/sycl-post-link/SYCLDeviceLibReqMask.cpp:203]: Hi, @gmlueck 
Long time ago, clang introduced one bug and would lead to a fp related flag missing in sycl device code and would lead to perf drop for sycl programs. However, libsycl-fallback-cassert.spv did include this flag and when linked with user's device code, it would save the perf by happy coincidence. This bug has nothing to do with libdevice and has been fixed.

Thanks very much.

**Review by gmlueck — APPROVED** [review]

File: clang/test/Driver/sycl-device-lib.cpp [hunk 2]
```diff
@@ -118,12 +100,6 @@
...
-// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown" "-inputs={{.*}}libsycl-fallback-cstring.o" "-outputs={{.*}}libsycl-fallback-cstring-{{.*}}.o" "-unbundle"
-// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown" "-inputs={{.*}}libsycl-fallback-complex.o" "-outputs={{.*}}libsycl-fallback-complex-{{.*}}.o" "-unbundle"
-// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown" "-inputs={{.*}}libsycl-fallback-complex-fp64.o" "-outputs={{.*}}libsycl-fallback-complex-fp64-{{.*}}.o" "-unbundle"
-// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown" "-inputs={{.*}}libsycl-fallback-cmath.o" "-outputs={{.*}}libsycl-fallback-cmath-{{.*}}.o" "-unbundle"
-// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown" "-inputs={{.*}}libsycl-fallback-cmath-fp64.o" "-outputs={{.*}}libsycl-fallback-cmath-fp64-{{.*}}.o" "-unbundle"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-device-lib.cpp:126]: Can you add a check that the fallback device libs are `-NOT` added under the expected circumstances?

**Comment by jinge90** [comment]
Hi, @tfzhu 
Error: Path does not exist /localdisk2/github/_work/llvm/llvm/build/results_l0_gen9_default.json
Error: Exit code 1 returned from process: file name '/localdisk2/github/bin.2.287.1/Runner.PluginHost', arguments 'action 
Is this an infrastructure issue?

Thanks very much.

File: clang/test/Driver/sycl-device-lib.cpp (hunk 2 above)
Comment (jinge90) [reply clang/test/Driver/sycl-device-lib.cpp:126]: Hi, @mdtoguchi 
I have added 2 tests for checking fallback devicelibs are not linked by default.

Thanks very much.

**Comment by tfzhu** [comment]
> 
looks like connection glitch, re-running

File: clang/test/Driver/sycl-device-lib.cpp (hunk 2 above)
Comment (jinge90) [reply clang/test/Driver/sycl-device-lib.cpp:126]: @mdtoguchi 
Could you help review if the test update meets your expectation?
Thanks very much.

**Review by AlexeySachkov — APPROVED** [review]
`sycl-post-link` part LGTM

**Comment by jinge90** [comment]
Hi, @intel/llvm-reviewers-runtime and @dm-vodopyanov 
Could you help review this patch?

Thanks very much.

File: libdevice/cmath_wrapper.cpp [hunk 3]
```diff
@@ -23,12 +23,10 @@ DEVICE_EXTERN_C
...
 
 DEVICE_EXTERN_C
-ldiv_t ldiv(long int x, long int y) { return __devicelib_ldiv(x, y); }
+ldiv_t ldiv(long x, long y) { return __devicelib_ldiv(x, y); }
 
 DEVICE_EXTERN_C
-lldiv_t lldiv(long long int x, long long int y) {
-  return __devicelib_lldiv(x, y);
-}
+lldiv_t lldiv(long long x, long long y) { return __devicelib_lldiv(x, y); }
```
Comment (s-kanaev) [inline libdevice/cmath_wrapper.cpp:30]: What's the purpose of these changes?

**Review by s-kanaev — COMMENTED** [review]
RT changes (program manager) look good.

Are there any tests to enable? If not, I reckon, there should be some tests added to verify online-linking works fine.

**Review by mdtoguchi — APPROVED** [review]

File: libdevice/cmath_wrapper.cpp (hunk 3 above)
Comment (jinge90) [reply libdevice/cmath_wrapper.cpp:30]: Just to align with math.h prototype, a minor NFC change.

Thanks very much.

**Review by s-kanaev — APPROVED** [review]
